homeassistant/components/feedreader/coordinator.py

```diff
@@ -0,0 +1,247 @@
+            "coordinator initi with url:%s scan_interval:%s", url, scan_interval
+        )
+
+    def _log_no_entries(self) -> None:
+        """Send no entries log at debug level."""
+        _LOGGER.debug("No new entries to be published in feed %s", self._url)
+
+    async def _async_update_data(self) -> dict:
+        """Update the feed and publish new entries to the event bus."""
+        last_entry_timestamp = await self.hass.async_add_executor_job(self._update)
+        if last_entry_timestamp:
+            self._storage.async_put_timestamp(self._feed_id, last_entry_timestamp)
+        return {}
+
+    def _update(self) -> struct_time | None:
+        """Update the feed and publish new entries to the event bus."""
+        _LOGGER.debug("Fetching new data from feed %s", self._url)
+        self._feed = feedparser.parse(
+            self._url,
+            etag=None if not self._feed else self._feed.get("etag"),
+            modified=None if not self._feed else self._feed.get("modified"),
+        )
+        if not self._feed:
+            _LOGGER.error("Error fetching feed data from %s", self._url)
+            return None
+        # The 'bozo' flag really only indicates that there was an issue
+        # during the initial parsing of the XML, but it doesn't indicate
+        # whether this is an unrecoverable error. In this case the
+        # feedparser lib is trying a less strict parsing approach.
+        # If an error is detected here, log warning message but continue
+        # processing the feed entries if present.
+        if self._feed.bozo != 0:
+            _LOGGER.warning(
+                "Possible issue parsing feed %s: %s",
+                self._url,
+                self._feed.bozo_exception,
+            )
+        # Using etag and modified, if there's no new data available,
+        # the entries list will be empty
+        _LOGGER.debug(
+            "%s entri(es) available in feed %s",
+            len(self._feed.entries),
+            self._url,
+        )
+        if not self._feed.entries:
+            self._log_no_entries()
+            return None
+
+        self._filter_entries()
+        self._publish_new_entries()
+
+        _LOGGER.debug("Fetch from feed %s completed", self._url)
+
+        if (
+            self._has_published_parsed or self._has_updated_parsed
+        ) and self._last_entry_timestamp:
+            return self._last_entry_timestamp
+
+        return None
+
+    def _filter_entries(self) -> None:
+        """Filter the entries provided and return the ones to keep."""
+        assert self._feed is not None
+        if len(self._feed.entries) > self._max_entries:
+            _LOGGER.debug(
+                "Processing only the first %s entries in feed %s",
+                self._max_entries,
+                self._url,
+            )
+            self._feed.entries = self._feed.entries[0 : self._max_entries]
+
+    def _update_and_fire_entry(self, entry: feedparser.FeedParserDict) -> None:
```

Comment:
This work should be done inside the event loop and not in a thread.

Reply:
b5a5211a048ca16cc4355fa2d3a6a87aeae67385